### PR TITLE
Added a typescript defs file

### LINKED ISSUE #3: TypeScript support
First of all love this lib! Neat. :100:

TypeScript people had `@types/classnames` - with this lib being a drop in replacement, should we fork that typings, or simply include it in this repo.

```ts
declare module 'clsx' {
	type ClassValue = string | number | ClassDictionary | ClassArray | undefined | null | boolean;

	interface ClassDictionary {
		[ id: string ]: any;
	}

	interface ClassArray extends Array<ClassValue> {
	}

	type ClassNamesFn = (...classes: ClassValue[]) => string;

	type ClassNamesExport = ClassNamesFn & { default: ClassNamesFn };

	const classNames: ClassNamesExport;

	export = classNames;
}
```
> sourced from: https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/classnames/index.d.ts

Not sure if you wan't to include this in your project as a `index.d.ts` in the root.

### PR DESCRIPTION
**Pull request by maraisr**
Not sure we need all the fancy things from that @types/classnames repo - but here is something nice and tight.

If you're not too keen; happy to amend this with a simple copy-paste from their one, and rename somethings.

resolves #3

### CODE REVIEW

File: index.d.ts [hunk 1]
```diff
@@ -0,0 +1,12 @@
...
+
+interface ClassDictionary {
+	[id: string]: any;
+}
+
+interface ClassArray extends Array<ClassValue> { }
+
+declare const clsx: (...classes: ClassValue[]) => string;
+
+export = clsx;
```
Comment (TrySound) [inline index.d.ts]: Why not normal `export default`? IMO esModuleInterop should be enabled by default.

Comment (kentcdodds) [reply index.d.ts]: I agree

File: index.d.ts [hunk 2]
```diff
@@ -0,0 +1,12 @@
+export type ClassValue = ClassArray | ClassDictionary | string | number | null | boolean;
+
+interface ClassDictionary {
```
Comment (kentcdodds) [inline index.d.ts]: We need to export this as well

File: index.d.ts [hunk 3]
```diff
@@ -0,0 +1,12 @@
...
+	[id: string]: any;
+}
+
+interface ClassArray extends Array<ClassValue> { }
```
Comment (kentcdodds) [inline index.d.ts]: Let's export this too

**Comment by lukeed** [comment]
Aye, I'll take care of properly distributing the types to npm – that's about all I can contribute here. 

As per Twitter thread, I can't really weigh in on the right/wrong/better ways for typings, so once you all come to an agreement I'll merge.

---

I will say that multiple of my repos have been PR'd with `export = lib`. @marvinhagemeister shared [this link](https://www.typescriptlang.org/docs/handbook/modules.html#export--and-import--require) with me about it.

**Comment by kentcdodds** [comment]
I'm pretty new to TS, but from what I understand `export =` is an old way to do things that pre-dates the ESModules specification. In my own version of these definitions, I'm using `export default` and that's working well :+1: (plus `export {ClassArray, ClassDictionary, ClassValue}`)

**Comment by lukeed** [comment]
Ah, then I suppose the above was specific to `kleur` since it's Node/CommonJS only & can be required as default or partially.

**Comment by lukeed** [comment]
Is this final form?

```ts
export type ClassValue = ClassArray | ClassDictionary | string | number | null | boolean;

export interface ClassDictionary {
	[id: string]: any;
}

export interface ClassArray extends Array<ClassValue> { }

declare const clsx: (...classes: ClassValue[]) => string;

export default clsx;
```

**Comment by kentcdodds** [comment]
Yep, I think that'll do it!

**Comment by marvinhagemeister** [comment]
yup, the `export = foo` way is only needed when the library doesn't use proper exports and uses plain `module.exports = foo` instead. The only exception is when the special `__esModule` property is defined on the object exported by `module.exports`.

In the case for `clsx` the proper way to do it is via a `export default` expression as already suggested 👍

**Comment by kentcdodds** [comment]
Thanks for the help @marvinhagemeister!

**Comment by lukeed** [comment]
Thank you @marvinhagemeister, makes sense~!

And thank you all, especially @maraisr, for the types. Since they are so simple & small, I'll just add them here after all. Patch incoming~

**Comment by kentcdodds** [comment]
👏

**Comment by lukeed** [comment]
Released in `1.0.1` 🙌

**Pull request merged**